### ISSUE TITLE
Frontend does not work in GRPC mode

### ISSUE THREAD

**Issue by siavashs** [grafana/loki#2533]
**Describe the bug**
Frontend fails to work in GRPC mode as described in the documentation: https://grafana.com/docs/loki/latest/configuration/query-frontend/
Also the documentation does not mention the pros/cons of using GRPC mode if any.

**To Reproduce**
1. Install Loki using the Helm chart (subcharts for each component)
2. Remove the following snippet from config:
    ```yaml
      frontend:
        downstream_url: http://loki-querier:3100
    ```
3. Add the following snippet to config:
    ```yaml
      frontend_worker:
        frontend_address: loki-query-frontend:9095
    ```
4. Add the GRPC port:
    ```yaml
    extraPorts:
      - port: 9095
        protocol: TCP
        name: grpc
        targetPort: grpc
    ```

After the above changes, `frontend` fails and responds with `499` errors, the pod never enters ready state and stays in `Running 0/1` status forever.

**Expected behavior**
Frontend should work in GRPC mode as described in the documentation.

**Environment:**
 - Infrastructure: Kubernetes (minikube)
 - Deployment tool: helm

**Screenshots, Promtail config, or terminal output**
```
> kubectl logs -f loki-query-frontend-0
level=info ts=2020-08-21T12:33:00.856420607Z caller=main.go:124 msg="Starting Loki" version="(version=1.6.0, branch=HEAD, revision=6e3e2fa39)"
level=info ts=2020-08-21T12:33:00.856914192Z caller=server.go:194 http=[::]:3100 grpc=[::]:9095 msg="server listening on addresses"
level=debug ts=2020-08-21T12:33:00.857130634Z caller=modules.go:285 msg="initializing query frontend" config="{Config:{MaxOutstandingPerTenant:100 CompressResponses:false DownstreamURL: LogQueriesLongerThan:0s} TailProxyUrl:}"
level=debug ts=2020-08-21T12:33:00.857307066Z caller=modules.go:290 msg="initializing query range tripperware" config="{Config:{SplitQueriesByInterval:15m0s SplitQueriesByDay:false AlignQueriesWithStep:true ResultsCacheConfig:{CacheConfig:{EnableFifoCache:false DefaultValidity:24h0m0s Background:{WriteBackGoroutines:10 WriteBackBuffer:10000} Memcache:{Expiration:24h0m0s BatchSize:100 Parallelism:100} MemcacheClient:{Host:loki-memcached.default.svc.cluster.local Service:memcache Addresses: Timeout:100ms MaxIdleConns:16 UpdateInterval:1m0s ConsistentHash:true} Redis:{Endpoint: Timeout:100ms Expiration:0s MaxIdleConns:80 MaxActiveConns:0 Password: EnableTLS:false IdleTimeout:0s WaitOnPoolExhaustion:false MaxConnLifetime:0s} Fifocache:{MaxSizeBytes: MaxSizeItems:0 Validity:0s DeprecatedSize:0} Prefix:frontend. Cache:<nil>} LegacyMaxCacheFreshness:0s} CacheResults:true MaxRetries:5 ShardedQueries:false}}" limits="{IngestionRateStrategy:local IngestionRateMB:4 IngestionBurstSizeMB:6 MaxLabelNameLength:1024 MaxLabelValueLength:2048 MaxLabelNamesPerSeries:30 RejectOldSamples:true RejectOldSamplesMaxAge:168h0m0s CreationGracePeriod:10m0s EnforceMetricName:false MaxLineSize:0B MaxLocalStreamsPerUser:10000 MaxGlobalStreamsPerUser:0 MaxChunksPerQuery:2000000 MaxQueryLength:0s MaxQueryParallelism:14 CardinalityLimit:100000 MaxStreamsMatchersPerQuery:1000 MaxConcurrentTailRequests:10 MaxEntriesLimitPerQuery:5000 MaxCacheFreshness:1m0s QuerySplitDuration:0s PerTenantOverrideConfig: PerTenantOverridePeriod:10s}"
level=debug ts=2020-08-21T12:33:00.859097513Z caller=module_service.go:48 msg="module waiting for initialization" module=query-frontend waiting_for=server
level=info ts=2020-08-21T12:33:00.859241595Z caller=module_service.go:58 msg=initialising module=server
level=info ts=2020-08-21T12:33:00.859324406Z caller=module_service.go:58 msg=initialising module=query-frontend
level=info ts=2020-08-21T12:33:00.859352246Z caller=loki.go:210 msg="Loki started"
level=debug ts=2020-08-21T12:33:49.019516329Z caller=logging.go:57 traceID=3555736b82334f03 msg="GET /ready (499) 1.000445238s"
level=debug ts=2020-08-21T12:33:52.81512663Z caller=logging.go:57 traceID=794407dc6a2977b9 msg="GET /ready (499) 999.864162ms"
level=debug ts=2020-08-21T12:33:59.019106627Z caller=logging.go:57 traceID=5d1aeb9e79d08601 msg="GET /ready (499) 999.794318ms"
level=debug ts=2020-08-21T12:34:02.815746162Z caller=logging.go:57 traceID=186f970f15a1f417 msg="GET /ready (499) 999.945647ms"
level=debug ts=2020-08-21T12:34:09.0192503Z caller=logging.go:57 traceID=24f2ae77102386c2 msg="GET /ready (499) 1.000012156s"
level=debug ts=2020-08-21T12:34:12.816244538Z caller=logging.go:57 traceID=3488173e7c312590 msg="GET /ready (499) 1.000461524s"
level=info ts=2020-08-21T12:34:12.985469038Z caller=signals.go:55 msg="=== received SIGINT/SIGTERM ===\n*** exiting"
level=debug ts=2020-08-21T12:34:12.9866169Z caller=module_service.go:80 msg=stopping module=query-frontend
level=debug ts=2020-08-21T12:34:12.986715311Z caller=module_service.go:103 msg="module waiting for" module=server waiting_for=query-frontend
level=debug ts=2020-08-21T12:34:19.019070601Z caller=logging.go:57 traceID=3d2e6691c9d6d86 msg="GET /ready (499) 999.676035ms"
level=debug ts=2020-08-21T12:34:29.01943387Z caller=logging.go:57 traceID=76de1d423340eebb msg="GET /ready (499) 999.46972ms"
level=debug ts=2020-08-21T12:34:39.019380354Z caller=logging.go:57 traceID=1ec40967bb8eaa06 msg="GET /ready (499) 999.237459ms"
level=debug ts=2020-08-21T12:34:49.019354728Z caller=logging.go:57 traceID=6a0f5dd3b34e6cd8 msg="GET /ready (499) 999.73175ms"
level=debug ts=2020-08-21T12:34:59.019431504Z caller=logging.go:57 traceID=7f317511b2fdaf73 msg="GET /ready (499) 999.857887ms"
level=debug ts=2020-08-21T12:35:09.019025598Z caller=logging.go:57 traceID=7065c4d39dcea7ca msg="GET /ready (499) 999.421858ms"
level=debug ts=2020-08-21T12:35:19.019169602Z caller=logging.go:57 traceID=294ddf725a3f2979 msg="GET /ready (499) 999.770351ms"
level=debug ts=2020-08-21T12:35:29.019201705Z caller=logging.go:57 traceID=2828ca3d74c4ad12 msg="GET /ready (499) 1.000245679s"
level=debug ts=2020-08-21T12:35:39.027897192Z caller=logging.go:57 traceID=22cc346bc23c6704 msg="GET /ready (499) 1.008792113s"
level=debug ts=2020-08-21T12:35:49.019824101Z caller=logging.go:57 traceID=297d5abdb4e094fa msg="GET /ready (499) 999.941868ms"
...
```

**Comment by tivvit** [comment]
I am also experiencing issues with frontend in pull mode. I am working on some more comprehensive report

**Comment by siavashs** [comment]
I injected many break points into the loki server, frontend and healthcheck code but none captured the call to `/` or `/metrics`, I guess that root cause is somewhere in cortex codebase.

**Comment by slim-bean** [comment]
This should have been fixed in #2610
